### DISCUSSION TITLE
Checking for package vulnerabilities without running restore

### DISCUSSION THREAD

**Discussion by trampster** [NuGet/Home#12087] [Q&A] [answered]
`dotnet list package --vulnerable --include-transitive` requires you to have previous run restore.

But nuget packages can run powershell scripts on install, which means by the time dotnet list tells you about the vulnerability it is already to late.

How can we check our project has no vulnerable packages without running restore and potentially infecting our machine?

**Comment by zivkan** [comment, marked as answer]
`dotnet list package` only works on projects using PackageReference, not packages.config. In fact, the `dotnet` CLI only officially works on SDK style projects. Visual Studio's older style .NET Framework projects, or other Visual Studio projects, may not work with the `dotnet` CLI if they use MSBuild `<import` statements for props/targets files that are only available in Visual Studio.

Anyway, as documented in the [migrate from packages.config to PackageReference documentation](https://docs.microsoft.com/en-us/nuget/consume-packages/migrate-packages-config-to-package-reference), with PackageReference projects, `install.ps1` is no longer executed on install. Therefore, any project that `dotnet list package` supports will never execute a package's install.ps1.

Additionally, I want to clarify that restore is not the same operation as install. Install is an action where a project that did not previously reference a package will "reference" the package at the end of the operation. In other words, it's the "install" button in Visual Studio's Package Manager UI, or the `Install-Package` cmdlet in Visual Studio's Package Manager Console. However, restore is not adding any new PackageReference items to the project (or packages to packages.config), it is making sure that the packages that already existed in the project/packages.config file are available on disk, so that the rest of the build system can access it.  Therefore, even with packages.config, restore is a "safe" operation, as restore will never execute install.ps1.

Packages can still bundle MSBuild props/targets, but command line restore will never execute MSBuild targets shipped in packages. Therefore, if you check for vulnerabilities after restore, but before build, there is no opportunity for a package to execute a malicious payload.

**Reply by trampster** [reply]
Thanks for that great answer.
